File: lib/rgfatools/multiplication.rb
#
# Methods which edit the graph components without traversal
#
module RGFATools::Multiplication

  # Allowed values for the links_distribution_policy option
  LINKS_DISTRIBUTION_POLICY = [:off, :auto, :equal, :E, :B]

  # @overload multiply(segment, factor, copy_names: :lowcase, distribute: :auto, conserve_components: true, origin_tag: :or)
  # Create multiple copies of a segment.
  #
  # Complements the multiply method of gfatools with additional functionality.
  # These extensions are used only after #enable_extensions is called on the
  # RGFA object. After that, you may still call the original method
  # using #multiply_without_rgfatools.
  #
  # For more information on the additional functionality, see
  # #multiply_extended.
  #
  # @return [RGFA] self
  def multiply_with_rgfatools(segment, factor,
                       copy_names: :lowcase,
                       distribute: :auto,
                       conserve_components: true,
                       origin_tag: :or)
    if !@extensions_enabled
      return multiply_without_rgfatools(segment, factor,
                       copy_names: copy_names,
                       conserve_components: conserve_components)
    else
      multiply_extended(segment, factor,
                       copy_names: copy_names,
                       distribute: distribute,
                       conserve_components: conserve_components,
                       origin_tag: origin_tag)
    end
  end

  # Create multiple copies of a segment.
  #
  # Complements the multiply method of gfatools with additional functionality.
  # To always run the additional functionality when multiply is called,
  # use RGFA#enable_extensions.
  #
  # @!macro [new] copynames_text
  #
  #   <b>Automatic computation of the copy names:</b>
  #
  #   - First, itis checked if the name of the original segment ends with a
  #     relevant
  #     string, i.e. a lower case letter (for +:lowcase+), an upper case letter
  #     (for +:upcase+), a digit (for +:number+), or the string +"_copy"+
  #     plus one or more optional digits (for +:copy+).
  #   - If so, it is assumed, it was already a copy, and it is not
  #     altered.
  #   - If not, then +a+ (for +:lowcase+), +A+ (for +:upcase+), +1+ (for
  #     +:number+), +_copy+ (for +:copy+) is appended to the string.
  #   - Then, in all
  #     cases, next (*) is called on the string, until a valid, non-existant
  #     name is found for each of the segment copies
  #   - (*) = except for +:copy+, where
  #     for the first copy no digit is present, but for the following is,
  #     i.e. the segment names will be +:copy+, +:copy2+, +:copy3+, etc.
  # - Can be overridden, by providing an array of copy names.
  #
  # @!macro [new] ldp_text
  #
  #   <b>Links distribution policy</b>
  #
  #   Depending on the value of the option +distribute+, an end
  #   is eventually selected for distribution of the links.
  #
  #   - +:off+: no distribution performed
  #   - +:E+: links of the E end are distributed
  #   - +:B+: links of the B end are distributed
  #   - +:equal+: select an end for which the number of links is equal to
  #     +factor+, if any; if both, then the E end is selected
  #   - +:auto+: automatically select E or B, trying to maximize the number of
  #     links which can be deleted
  #
  # @param [Integer] factor multiplication factor; if 0, delete the segment;
  #   if 1; do nothing; if > 1; number of copies to create
  # @!macro [new] segment_param
  #   @param segment [String, RGFA::Line::Segment] segment name or instance
  # @param [:lowcase, :upcase, :number, :copy, Array<String>] copy_names
  #   <i>(Defaults to: +:lowcase+)</i>
  #   Array of names for the copies of the segment,
  #   or a symbol, which defines a system to compute the names from the name of
  #   the original segment. See "Automatic computation of the copy names".
  # @!macro [new] conserve_components
  #   @param [Boolean] conserve_components <i>(Defaults to: +true+)</i>
  #     If factor == 0 (i.e. deletion), delete segment only if
  #     #cut_segment?(segment) is +false+ (see RGFA API).
  # @!macro [new] ldp_param
  #   @param distribute
  #     [RGFATools::Multiplication::LINKS_DISTRIBUTION_POLICY]
  #     <i>(Defaults to: +:auto+)</i>
  #     Determines if and for which end of the segment, links are distributed
  #     among the copies. See "Links distribution policy".
  # @!macro [new] origin_tag
  #   @param origin_tag [Symbol] <i>(Defaults to: +:or+)</i>
  #     Name of the custom tag to use for storing origin information.
  #
  # @return [RGFA] self
  def multiply_extended(segment, factor,
                       copy_names: :lowcase,
                       distribute: :auto,
                       conserve_components: true,
                       origin_tag: :or)
    s, sn = segment_and_segment_name(segment)
    s.set(origin_tag, sn) if !s.get(origin_tag)
    copy_names = compute_copy_names(copy_names, sn, factor)
    multiply_without_rgfatools(sn, factor,
                               copy_names: copy_names,
                               conserve_components: conserve_components)
    distribute_links(distribute, sn, copy_names, factor)
    return self
  end

  private

  Redefined = [:multiply]

  def select_distribute_end(links_distribution_policy, segment_name, factor)
    accepted = RGFATools::Multiplication::LINKS_DISTRIBUTION_POLICY
    if !accepted.include?(links_distribution_policy)
      raise "Unknown links distribution policy #{links_distribution_policy}, "+
        "accepted values are: "+
        accepted.inspect
    end
    return nil if links_distribution_policy == :off
    if [:B, :E].include?(links_distribution_policy)
      return links_distribution_policy
    end
    s = segment(segment_name)
    esize = s.dovetails(:R).size
    bsize = s.dovetails(:L).size
    auto_select_distribute_end(factor, bsize, esize,
                               links_distribution_policy == :equal)
  end

  # (keep separate for testing)
  def auto_select_distribute_end(factor, bsize, esize, equal_only)
    if esize == factor
      return :E
    elsif bsize == factor
      return :B
    elsif equal_only
      return nil
    elsif esize < 2
      return (bsize < 2) ? nil : :B
    elsif bsize < 2
      return :E
    elsif esize < factor
      return ((bsize <= esize) ? :E :
        ((bsize < factor) ? :B : :E))
    elsif bsize < factor
      return :B
    else
      return ((bsize <= esize) ? :B : :E)
    end
  end

  def distribute_links(links_distribution_policy, segment_name,
                       copy_names, factor)
    return if factor < 2
    end_type = select_distribute_end(links_distribution_policy,
                                     segment_name, factor)
    return nil if end_type.nil?
    et_links = segment(segment_name).dovetails(end_type == :B ? :L : :R)
    diff = [et_links.size - factor, 0].max
    links_signatures = et_links.map do |l|
      l.other_end([segment_name, end_type]).join
    end
    ([segment_name]+copy_names).each_with_index do |sn, i|
      segment(sn).dovetails(end_type == :B ? :L : :R).each do |l|
        l_sig = l.other_end([sn, end_type]).join
        to_save = links_signatures[i..i+diff].to_a
        l.disconnect! unless to_save.include?(l_sig)
      end
    end
  end

  def segment_and_segment_name(segment_or_segment_name)
    if segment_or_segment_name.kind_of?(RGFA::Line)
      s = segment_or_segment_name
      sn = segment_or_segment_name.name
    else
      sn = segment_or_segment_name.to_sym
      s = segment(sn)
    end
    return s, sn
  end

end

File: gfapy/graph_operations/multiplication.py
import gfapy
import re

class Multiplication:

  def multiply(self, segment, factor, copy_names = None,
               conserve_components = True, distribute = None,
               track_origin = False, origin_tag="or", extended = False):
    """Multiply a segment by a given factor.

    The multiplication operation is implemented as described in
    Gonnella and Kurtz (2016).

    Parameters:
      segment (Line, str): the segment to multiply
      factor (int): the multiplication factor; if 0, the segment is
        deleted; if 1, nothing is done; if > 1, the multiplication
        is performed
      copy_names (list, None): an optional list of strings, the names
        of the copies which will result from the multiplication;
        the length of this list must be equal to factor - 1; if no
        list is specified, the names are computed automatically, adding
        (or incrementing) an integer as suffix to the segment name,
        until enough non-previously used names are found
      conserve_components (bool): if True, the removal of segments
        in the case where factor == 0 is only done if it does not
        split an existing connected component (thereby only dovetail
        overlaps are considered)
      extended : if True, then dovetail distribution and track origin
        are turned on by default
      distribute (str, None) : select an end for which the dovetail
        overlaps are distributed (see Gonnella and Kurtz, 2016); if ``auto``
        (the default if extended is set), an end is selected automatically,
        trying to maximize the number of links which can be deleted; if ``off``
        (the default if extended is not set), no distribution is performed; if
        ``L`` or ``R``, links of the specified end are distributed; if
        ``equal``, an end is selected (if any), for which the number of links
        is equal to the factor (if none, links are not distributed; if both,
        then ``R`` is used)
      track_origin (bool): if True, the name of the original segment (or
        the content of its own origin tag, if any) is stored
        in a tag in the copies (default: False)
      origin_tag (str): the tag where to store the origin information,
        if track_origin is set (default: ``or``)
    """
    if extended:
      if distribute == None:
        distribute = "auto"
      track_origin = True
    if factor < 0:
      raise gfapy.ArgumentError("Mulitiplication factor must be >= 0"+
          " ({} found)".format(factor))
    elif factor == 0:
      if conserve_components and factor == 1 and self.is_cut_segment(segment):
        return self
      else:
        self.rm(segment)
        return self
    elif factor == 1:
      return self
    else:
      s, sn = self._segment_and_segment_name(segment)
      if track_origin and not s.get(origin_tag):
        s.set(origin_tag, sn)
      self.__divide_segment_and_connection_counts(s, factor)
      if copy_names is None:
        copy_names = self._compute_copy_names(sn, factor)
      for cn in copy_names:
        self.__clone_segment_and_connections(s, cn)
      if distribute:
        self._distribute_links(distribute, sn, copy_names, factor)
      return self

  def _compute_copy_names(self, segment_name, factor):
    assert factor >= 2
    retval = []
    first = 2
    m = re.search(r'(.*)\*(\d+)',segment_name)
    if m:
      segment_name = m.groups()[0]
      i = int(m.groups()[1])
    offset = 0
    for i in range(first,factor+first-1):
      name = "{}*{}".format(segment_name, i+offset)
      while name in self.names:
        offset+=1
        name = "{}*{}".format(segment_name, i+offset)
      retval.append(name)
    return retval

  def __divide_counts(self, gfa_line, factor):
    for count_tag in ["KC", "RC", "FC"]:
      if count_tag in gfa_line.tagnames:
        gfa_line.set(count_tag, gfa_line.get(count_tag) // factor)

  def __divide_segment_and_connection_counts(self, segment, factor):
    self.__divide_counts(segment, factor)
    processed_circulars = set()
    for l in segment.dovetails + segment.containments:
      if l.is_circular():
        if l not in processed_circulars:
          self.__divide_counts(l, factor)
          processed_circulars.add(l)
      else:
        self.__divide_counts(l, factor)

  def __clone_segment_and_connections(self, segment, clone_name):
    cpy = segment.clone()
    cpy.name = clone_name
    cpy.connect(self)
    for l in segment.dovetails + segment.containments:
      lc = l.clone()
      if lc.from_segment == segment.name:
        lc.from_segment = clone_name
      if lc.to_segment == segment.name:
        lc.to_segment = clone_name
      lc.connect(self)

  LINKS_DISTRIBUTION_POLICY = ["off", "auto", "equal", "L", "R"]
  '''Allowed values for the links_distribution_policy option'''

  def _select_distribute_end(self, links_distribution_policy,
                             segment_name, factor):
    if links_distribution_policy not in self.LINKS_DISTRIBUTION_POLICY:
      raise gfapy.ArgumentError("Unknown links distribution policy {}\n".format(\
          links_distribution_policy)+"accepted values are: {}".format(\
          ", ".join(self.LINKS_DISTRIBUTION_POLICY)))
    if links_distribution_policy == "off":
      return None
    if links_distribution_policy in ["L", "R"]:
      return links_distribution_policy
    else:
      s = self.segment(segment_name)
      esize = len(s.dovetails_of_end("R"))
      bsize = len(s.dovetails_of_end("L"))
      return self._auto_select_distribute_end(factor, bsize, esize,
                                       links_distribution_policy == "equal")

  # (keep separate for testing)
  # @tested_in unit_multiplication
  @staticmethod
  def _auto_select_distribute_end(factor, bsize, esize, equal_only):
    if esize == factor:
      return "R"
    elif bsize == factor:
      return "L"
    elif equal_only:
      return None
    elif esize < 2:
      if bsize < 2:
        return None
      else:
        return "L"
    elif bsize < 2:
      return "R"
    elif esize < factor:
      if bsize <= esize:
        return "R"
      elif bsize < factor:
        return "L"
      else:
        return "R"
    elif bsize < factor:
      return "L"
    elif bsize <= esize:
      return "L"
    else:
      return "R"

  def _distribute_links(self, links_distribution_policy, segment_name,
                        copy_names, factor):
    if factor < 2:
      return
    end_type = self._select_distribute_end(links_distribution_policy,
                                           segment_name, factor)
    if end_type is None:
      return
    et_links = self.segment(segment_name).dovetails_of_end(end_type)
    diff = max([len(et_links)-factor, 0])
    links_signatures = list([repr(l.other_end(gfapy.SegmentEnd(segment_name, \
                          end_type))) for l in et_links])
    for i, sn in enumerate([segment_name]+copy_names):
      to_keep = links_signatures[i:i+diff+1]
      links = self.segment(sn).dovetails_of_end(end_type).copy()
      for l in links:
        l_sig = repr(l.other_end(gfapy.SegmentEnd(sn, end_type)))
        if l_sig not in to_keep:
          l.disconnect()

  def _segment_and_segment_name(self, segment_or_segment_name):
    if isinstance(segment_or_segment_name, gfapy.Line):
      s = segment_or_segment_name
      sn = segment_or_segment_name.name
    else:
      s = self.segment(segment_or_segment_name)
      sn = segment_or_segment_name
    return s, sn
